### ISSUE TITLE
RFC (Vue): Support usage in vue based web components with defineCustomElement

### ISSUE THREAD

**Issue by cimchd** [urql-graphql/urql#2879]
## Summary
We have a micro app architecture relying on web components build with Vue's [defineCustomElements](https://vuejs.org/guide/extras/web-components.html#building-custom-elements-with-vue) function.
Therefor we have no app instance an no way to add urql as a plugin or use provide/inject.

A workaround would be of course not to use defineCustomelement and create the web components manually with an app instance, but we want to use defineCustomElement and reduce the manual work per web component (we will have probably 100+ web components in our mono repo).

## Proposed Solution

To enable this use case there are 2 possible changes that can achieve this:

**1. Add an optional client parameter to useClientHandle**
We could then use this parameter instead of useClient to get the client:
```ts
export function useClientHandle(clientArg?: Client): ClientHandle {
  const client = clientArg ?? useClient();
  //[...]
}
```

**2. Export all callUse\* functions**
If callUseQuery, callUseMutation, etc. functions are exported, we could build our own useClientHandle function based on those functions.

Solution 1 is the best solution if this should be an officially supported use case by `@urql/vue`. If not then solution 2 would allow a manual reuse of the vue functions for building a custom workaround. Both changes can also be implemented together of course.

If one (or both) of the proposed changes are welcome, I can provide a PR.

**Comment by kitten** [comment]
So, if I'm reading this correctly it does say,

> The Provide / Inject API and its Composition API equivalent also work between Vue-defined custom elements.

This is basically our mechanism to enforce that the Provide API is used to deliver a client to these APIs in a targeted manner. Bypassing this is of course possible — although, I will rule out option two here for you, sorry, it's simply not something we're keen on doing (a library, in our view, is not a set of tools to recompose said library, to us; this would diverge different usecases and happy paths too far between users)

Option 1 — providing a client ref or a client directly is of course possible. But, I want to make sure that this is necessary.

I haven't tested this myself yet, but could you validate for me please whether using `provideClient` in the same or a wrapping custom element that uses the query APIs really doesn't work? I think the docs imply it should 🤔 https://github.com/urql-graphql/urql/blob/9abce5d7b1f125d9edd666994e15ca8ae0dec7a6/packages/vue-urql/src/useClient.ts#L4

**Comment by cimchd** [comment]
Thanks for your reply :-)
The problem with web components in vue is, that we don't have a parent/child component. We cannot provide and inject the same variable in the same component.

> I haven't tested this myself yet, but could you validate for me please whether using provideClient in the same or a wrapping custom element that uses the query APIs really doesn't work? I think the docs imply it should 🤔

I tested it. It works if I have a parent and a child component. But it doesn't work if I use provideClient in the same component. The behaviour is the same wether I use plain vue or web components with defineCustomElements.

For our use case I need something like this to work in a single component:

```vue
<script setup lang="ts">
import { useQuery, provideClient, createClient } from '@urql/vue';
import gql from 'graphql-tag';

const client = createClient({  url: `path/to/graphql` });
provideClient(client);

const { data } = useQuery({
  query: gql`
    query Customers {
      customers {
        id
        name
      }
    }
  `,
});
</script>

<template>
    <div>{{ data }}</div>
</template>
```
Currently we use vue-apollo in our architecture and it works. But we want to switch to urql. The provideApolloClient method from vue-apollo solved this issue by using a cache object instead of provide/inject: https://github.com/vuejs/apollo/blob/372892855d76622128ac560e8fadc689c50675bc/packages/vue-apollo-composable/src/useApolloClient.ts#L85

If provideClient stays the way it is, an extra client parameter to useClientHandle would allow us to build something like this for our project:
```ts
const { useQuery, useMutation, useSubscription, client } = createUrqlClient(`path/to/graphql` )
```
Where createUrqlClient is a wrapper for creating the client and returns the result of `useClientHandle`.

**Comment by kitten** [comment]
Sorry for forgetting about this!
I pushed a fix for this here: #3018. Very simple solution, but it'll allow for what you posted above.
